File: java/src/day07/CamelCards.java
package day07;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.*;

import static java.util.Map.entry;

public class CamelCards {

    private static final Map<String, Integer> handsBids = new HashMap<>();

    public static void main(String[] args) {
        try {
            getFile("inputs/day07/input.txt");
            System.out.println("Part 1: " + part1());
            System.out.println("Part 2: " + part2());
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static long part1() {
        List<Pair> rankHand = new ArrayList<>();
        for (String hand : handsBids.keySet()) {
            Map<Character, Integer> cards = getCountCards(hand);
            Integer typeOfHand = rankHand(cards);
            rankHand.add(new Pair(hand, typeOfHand));
        }
        rankHand.sort(Pair::sortPairs);
        return calculateWinnings(rankHand);
    }

    private static long part2() {
        List<Pair> rankHand = new ArrayList<>();
        for (String hand : handsBids.keySet()) {
            Map<Character, Integer> cards = getCountCardsWithJokers(hand);
            Integer typeOfHand = rankHand(cards);
            rankHand.add(new Pair(hand, typeOfHand));
        }
        rankHand.sort(Pair::sortPairs2);
        return calculateWinnings(rankHand);
    }

    private static long calculateWinnings(List<Pair> rankHand) {
        long result = 0;
        for (int i = 0; i < rankHand.size(); i++) {
            result += (long) handsBids.get(rankHand.get(i).hand) * (i + 1);
        }
        return result;
    }

    private static int rankHand(Map<Character, Integer> cards) {
        int combination = cards.values().size();
        // combination zero part2
        if (combination == 1 || combination == 0) {
            // Five of a kind
            return 7;
        } else if (combination == 2) {
            for (int value : cards.values()) {
                if (value == 4) {
                    // Four of a kind
                    return 6;
                }
            }
            // Full house
            return 5;
        } else if (combination == 3) {
            for (int value : cards.values()) {
                if (value == 3) {
                    // Three of a kind
                    return 4;
                }
            }
            // Two pair
            return 3;
        } else if (combination == 4) {
            // One pair
            return 2;
        }
        return 1;
    }

    private static Map<Character, Integer> getCountCards(String hand) {
        Map<Character, Integer> cards = new HashMap<>();

        for (char card : hand.toCharArray()) {
            cards.merge(card, 1, Integer::sum);
        }
        return cards;
    }

    private static Map<Character, Integer> getCountCardsWithJokers(String hand) {
        Map<Character, Integer> cards = new HashMap<>();
        List<Pair> sortedCards = new ArrayList<>();

        for (char card : hand.toCharArray()) {
            cards.merge(card, 1, Integer::sum);
        }
        for (Map.Entry<Character, Integer> card : cards.entrySet()) {
            sortedCards.add(new Pair(String.valueOf(card.getKey()), card.getValue()));
        }
        sortedCards.sort((par1, par2) -> par2.rank - par1.rank);
        Pair jokeR = null;
        for (int i = 0; i < sortedCards.size(); i++) {
            if (sortedCards.get(i).hand.equals("J")) {
                if (i == 0 && sortedCards.size() > 1) {
                    sortedCards.get(1).rank += sortedCards.get(i).rank;
                    jokeR = sortedCards.get(i);
                } else if (sortedCards.size() > 1) {
                    sortedCards.get(0).rank += sortedCards.get(i).rank;
                    jokeR = sortedCards.get(i);
                }
                break;
            }
        }
        if (jokeR != null) {
            sortedCards.remove(jokeR);
        }
        Map<Character, Integer> cardsWithJokers = new HashMap<>();
        for (Pair card : sortedCards) {
            cardsWithJokers.put(card.hand.charAt(0), card.rank);
        }
        return cardsWithJokers;
    }


    private static void getFile(String file) throws IOException {
        try (BufferedReader br = new BufferedReader(new FileReader(file))) {
            String line;

            while ((line = br.readLine()) != null) {
                String[] handBid = line.strip().split("\\s+");
                String hand = handBid[0];
                Integer bid = Integer.parseInt(handBid[1]);
                handsBids.put(hand, bid);
            }
        }
    }
}

class Pair {
    private static final Map<Character, Integer> valueOfCard = Map.ofEntries(
            entry('A', 13),
            entry('K', 12),
            entry('Q', 11),
            entry('J', 10),
            entry('T', 9),
            entry('9', 8),
            entry('8', 7),
            entry('7', 6),
            entry('6', 5),
            entry('5', 4),
            entry('4', 3),
            entry('3', 2),
            entry('2', 1)
    );

    private static final Map<Character, Integer> valueOfCard2 = Map.ofEntries(
            entry('A', 13),
            entry('K', 12),
            entry('Q', 11),
            entry('T', 9),
            entry('9', 8),
            entry('8', 7),
            entry('7', 6),
            entry('6', 5),
            entry('5', 4),
            entry('4', 3),
            entry('3', 2),
            entry('2', 1),
            entry('J', 0)
    );

    public String hand;
    public Integer rank;

    public Pair(String hand, Integer rank) {
        this.hand = hand;
        this.rank = rank;
    }

    public int sortPairs(Pair other) {
        if (Objects.equals(other.rank, this.rank)) {
            for (int i = 0; i < 5; i++) {
                char otherCard = other.hand.charAt(i);
                char thisCard = this.hand.charAt(i);
                if (Objects.equals(valueOfCard.get(otherCard), valueOfCard.get(thisCard))) {
                    continue;
                }
                return valueOfCard.get(thisCard) - valueOfCard.get(otherCard);
            }
        }
        return this.rank - other.rank;
    }

    public int sortPairs2(Pair other) {
        if (Objects.equals(other.rank, this.rank)) {
            for (int i = 0; i < 5; i++) {
                char otherCard = other.hand.charAt(i);
                char thisCard = this.hand.charAt(i);
                if (Objects.equals(valueOfCard2.get(otherCard), valueOfCard2.get(thisCard))) {
                    continue;
                }
                return valueOfCard2.get(thisCard) - valueOfCard2.get(otherCard);
            }
        }
        return this.rank - other.rank;
    }

    @Override
    public String toString() {
        return this.hand + ": " + this.rank;
    }
}

File: type-script/day07/camel-cards.ts
import * as fs from "fs";

const f: string = fs.readFileSync("../inputs/day07/input.txt", { encoding: "utf-8" });
const valueOfCardsPart1: Map<string, number> = new Map([
    ["A", 14],
    ["K", 13],
    ["Q", 12],
    ["J", 11],
    ["T", 10],
    ["9", 9],
    ["8", 8],
    ["7", 7],
    ["6", 6],
    ["5", 5],
    ["4", 4],
    ["3", 3],
    ["2", 2],
]);

const valueOfCardsPart2: Map<string, number> = new Map([
    ["A", 14],
    ["K", 13],
    ["Q", 12],
    ["T", 10],
    ["9", 9],
    ["8", 8],
    ["7", 7],
    ["6", 6],
    ["5", 5],
    ["4", 4],
    ["3", 3],
    ["2", 2],
    ["J", 1]
]);


function part1(file: string): number {
    const handsBids: Map<string, number> = getHandsFromFile(file);
    const hands: HandBid[] = [];
    for (let [k, v] of handsBids.entries()) {
        hands.push(new HandBid(k, v, valueOfCardsPart1));
    }
    hands.sort((a, b) => a.sortBy(b));
    const values: number[] = hands.map((hand, idx) => {
        return (idx + 1) * hand.bid;
    });

    return values.reduce((a, b) => a + b);
}

function part2(file: string): number {
    const handsBids: Map<string, number> = getHandsFromFile(file);
    const hands: HandBid[] = [];
    for (let [k, v] of handsBids.entries()) {
        hands.push(new HandBid(k, v, valueOfCardsPart2, true));
    }
    hands.sort((a, b) => a.sortBy(b));
    const values: number[] = hands.map((hand, idx) => {
        return (idx + 1) * hand.bid;
    });

    return values.reduce((a, b) => a + b);
}

function getHandsFromFile(file: string): Map<string, number> {
    const handsBids: Map<string, number> = new Map();
    const hands: string[][] = file.split("\n").map(x => x.trim().split(/\s+/));
    for (const hand of hands) {
        handsBids.set(hand[0], Number(hand[1]));
    }
    return handsBids;
}

class HandBid {
    private hand: string;
    bid: number;
    private valueOfCards: Map<string, number>;
    private countCards: Map<string, number>;
    private type: number;
    private jokers: boolean;

    constructor(k: string, v: number, valueOfCards: Map<string, number>, jokers: boolean = false) {
        this.hand = k;
        this.bid = v;
        this.jokers = jokers;
        this.valueOfCards = valueOfCards;
        this.countCards = this.getCountCards();
        this.type = this.typeOfHand();
    }

    private getCountCards(): Map<string, number> {
        const card: Map<string, number> = new Map();
        for (const char of this.hand) {
            if (card.get(char)) {
                card.set(char, (card.get(char) || 1) + 1);
                continue;
            }
            card.set(char, 1);
        }
        return card;
    }

    private typeOfHand(): number {
        let count: number = this.countCards.size;
        if (this.jokers) {
            this.correctValueForJokers(count);
        }

        // count === 0 for part2
        if (count === 1 || count === 0) {
            // Five of a kind
            return 6;
        } else if (count === 2) {
            for (let cnt of this.countCards.values()) {
                if (cnt === 4) {
                    // Four of a kind
                    return 5;
                }
            }
            // Full house
            return 4;
        } else if (count === 3) {
            for (let cnt of this.countCards.values()) {
                if (cnt === 3) {
                    // Three of a kind
                    return 3;
                }
            }
            // Two pair
            return 2
        } else if (count === 4) {
            // One pair
            return 1;
        }
        return 0;
    }

    private correctValueForJokers(count: number) {
        const valueOfJoker: number = this.countCards.get("J") || 0;
        if (valueOfJoker > 0) {
            this.countCards.delete("J");
            if (this.countCards.size > 0) {
                const [key, value]: [string, number] = [...this.countCards.entries()].reduce((a, e) => e[1] > a[1] ? e : a);
                this.countCards.set(key, value + valueOfJoker);
                count--;
            }
        }
        return count;
    }

    sortBy(other: HandBid): number {
        if (this.type !== other.type) {
            return this.type - other.type;
        }

        for (let i = 0; i < this.hand.length; i++) {
            if (this.hand[i] !== other.hand[i]) {
                return (this.valueOfCards.get(this.hand[i]) || 0)
                    - (this.valueOfCards.get(other.hand[i]) || 0);
            }
        }

        return 0;
    }
}

console.log("Part 1:", part1(f));
console.log("Part 2:", part2(f));
